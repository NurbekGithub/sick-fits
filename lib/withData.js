import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint, prodEndpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

export function dump(data) {
  return JSON.stringify(data, null, 2);
}

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            // 1. read the cartOpen value from the cache
            const { cartOpen } = cache.readQuery({ query: LOCAL_STATE_QUERY });
            // 2. write the cartOpen value to the cache
            const data = {
              data: { cartOpen: !cartOpen }
            };
            cache.writeData(data);
            return data;
          }
        }
      },
      defaults: {
        cartOpen: false
      }
    }
  });
}

export default withApollo(createClient);
