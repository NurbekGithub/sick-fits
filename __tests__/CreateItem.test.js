import { MockedProvider } from 'react-apollo/test-utils';
import waait from 'waait';
import { mount } from 'enzyme';
import CreateItem, { CREATE_ITEM_MUTATION } from '../components/CreateItem';
import toJSON from 'enzyme-to-json';
import Router from 'next/router'
import { fakeItem } from '../lib/testUtils';

const dogImage = 'https://dog.com/dog.png'
// mock the global fetch API
global.fetch = jest.fn().mockResolvedValue({
  json: () => ({
    secure_url: dogImage,
    eager: [{ secure_url: dogImage }]
  })
})

describe('<CreateItem />', () => {
  it('renders and matches snap', () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );
    const createItemForm = wrapper.find('form[data-test="createItem"]');
    expect(toJSON(createItemForm)).toMatchSnapshot();
  });
  it('uploads file when changed', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );
    const input = wrapper.find('input[type="file"]');
    input.simulate('change', { target: { files: ['fakefile.png'] } });
    await waait();
    const component = wrapper.find('CreateItem').instance();
    expect(component.state.image).toBe(dogImage);
    expect(component.state.largeImage).toBe(dogImage);
    expect(global.fetch).toHaveBeenCalled();
    global.fetch.mockReset();
  });
  it('handles state changing', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );
    const title = wrapper.find('input[name="title"]');
    const price = wrapper.find('input[name="price"]');
    const description = wrapper.find('textarea[name="description"]');
    title.simulate('change', {
      target: {
        name: 'title',
        value: 'doggy'
      }
    });
    price.simulate('change', {
      target: {
        name: 'price',
        value: 1212
      }
    });
    description.simulate('change', {
      target: {
        name: 'description',
        value: 'long long story'
      }
    });
    await waait();
    const component = wrapper.find('CreateItem').instance();
    expect(component.state.title).toBe('doggy');
    expect(component.state.description).toBe('long long story');
    expect(component.state.price).toBe(1212);
  });
  it('creates an item when the form is submitted', async () => {
    const item = fakeItem();
    const mocks = [
      {
        request: {
          query: CREATE_ITEM_MUTATION,
          variables: {
            title: item.title,
            description: item.description,
            price: item.price,
            image: '',
            largeImage: ''
          }
        },
        result: { data: { createItem: { ...item, id: 'abc', __typename: 'Item' } } }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <CreateItem />
      </MockedProvider>
    )
    // simulate filling the form;
    const title = wrapper.find('input[name="title"]');
    const price = wrapper.find('input[name="price"]');
    const description = wrapper.find('textarea[name="description"]');
    title.simulate('change', {
      target: {
        name: 'title',
        value: item.title
      }
    });
    price.simulate('change', {
      target: {
        name: 'price',
        value: item.price
      }
    });
    description.simulate('change', {
      target: {
        name: 'description',
        value: item.description
      }
    });
    // mock the router
    Router.router = { push: jest.fn() };
    wrapper.find('form').simulate('submit');
    await waait(50);
    expect(Router.router.push).toHaveBeenCalled();
    expect(Router.router.push).toHaveBeenCalledWith("/item?id=abc");
  })
});