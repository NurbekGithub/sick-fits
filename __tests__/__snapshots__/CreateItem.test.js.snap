// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem /> renders and matches snap 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  data-test="createItem"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <h2>
    Sell an item!
  </h2>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="image"
    >
      Image
      <input
        name="image"
        onChange={[Function]}
        placeholder="Image"
        required={true}
        type="file"
      />
    </label>
    <label
      htmlFor="title"
    >
      Title
      <input
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="price"
    >
      Price
      <input
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
        value={0}
      />
    </label>
    <label
      htmlFor="description"
    >
      Description
      <textarea
        name="description"
        onChange={[Function]}
        placeholder="Description"
        required={true}
        value=""
      />
    </label>
    <button>
      Submit
    </button>
  </fieldset>
</form>
`;
